File: python/can_place_flowers.py
class Solution:
    def canPlaceFlowers(self, flowerbed: list[int], n: int) -> bool:
        if n == 0:
            return True

        for i in range(len(flowerbed)):
            if flowerbed[i] == 1:
                if i > 0:
                    flowerbed[i - 1] = -1
                if i < len(flowerbed) - 1:
                    flowerbed[i + 1] = -1
        i: int = 0
        while i < len(flowerbed):
            if flowerbed[i] == 0:
                n -= 1
                if n == 0:
                    return True
                i += 2
                continue
            i += 1

        return False


def test_canPlaceFlowers_case_1():
    # arrange
    flowerbed: list[int] = [1, 0, 0, 0, 1]
    n: int = 1
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.canPlaceFlowers(flowerbed, n)

    # assert
    assert expected == actual


def test_canPlaceFlowers_case_2():
    # arrange
    flowerbed: list[int] = [1, 0, 1, 0, 1, 0, 1]
    n: int = 0
    expected: bool = True

    # act
    solution = Solution()
    actual = solution.canPlaceFlowers(flowerbed, n)

    # assert
    assert expected == actual

File: rust/src/code/can_place_flowers.rs
struct Solution {}

impl Solution {
    // NOTE: time complexity: O(n)
    pub fn can_place_flowers(mut flowerbed: Vec<i32>, mut n: i32) -> bool {
        if n == 0 {
            return true;
        }

        for i in 0..flowerbed.len() {
            if flowerbed[i] == 1 {
                if i > 0 {
                    flowerbed[i - 1] = -1;
                }
                if i < flowerbed.len() - 1 {
                    flowerbed[i + 1] = -1;
                }
            }
        }

        let mut i: usize = 0;
        while i < flowerbed.len() {
            if flowerbed[i] == 0 {
                n -= 1;
                if n == 0 {
                    return true;
                }
                i += 2;
                continue;
            }

            i += 1;
        }

        return if n > 0 { false } else { true };
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_can_place_flowers_case_1() {
        // arrange
        let flowerbed: Vec<i32> = vec![1, 0, 0, 0, 1];
        let n: i32 = 1;
        let expected: bool = true;

        // act
        let actual = Solution::can_place_flowers(flowerbed, n);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_can_place_flowers_case_2() {
        // arrange
        let flowerbed: Vec<i32> = vec![1, 0, 0, 0, 1];
        let n: i32 = 2;
        let expected: bool = false;

        // act
        let actual = Solution::can_place_flowers(flowerbed, n);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_can_place_flowers_case_3() {
        // arrange
        let flowerbed: Vec<i32> = vec![1, 0, 1, 0, 1, 0, 1];
        let n: i32 = 1;
        let expected: bool = false;

        // act
        let actual = Solution::can_place_flowers(flowerbed, n);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_can_place_flowers_case_4() {
        // arrange
        let flowerbed: Vec<i32> = vec![1, 0, 0, 0, 0, 1];
        let n: i32 = 2;
        let expected: bool = false;

        // act
        let actual = Solution::can_place_flowers(flowerbed, n);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_can_place_flowers_case_5() {
        // arrange
        let flowerbed: Vec<i32> = vec![1, 0, 0, 0, 1, 0, 0];
        let n: i32 = 2;
        let expected: bool = true;

        // act
        let actual = Solution::can_place_flowers(flowerbed, n);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_can_place_flowers_case_6() {
        // arrange
        let flowerbed: Vec<i32> = vec![0, 1, 0];
        let n: i32 = 1;
        let expected: bool = false;

        // act
        let actual = Solution::can_place_flowers(flowerbed, n);

        // assert
        assert_eq!(expected, actual);
    }

    #[test]
    fn test_can_place_flowers_case_7() {
        // arrange
        let flowerbed: Vec<i32> = vec![0, 1, 0, 1, 0, 1, 0, 0];
        let n: i32 = 1;
        let expected: bool = true;

        // act
        let actual = Solution::can_place_flowers(flowerbed, n);

        // assert
        assert_eq!(expected, actual);
    }
}
